src/Cli/dotnet/CommonOptions.cs

```diff
@@ -100,6 +116,81 @@ public static bool VerbosityIsDetailedOrDiagnostic(this VerbosityOptions verbosi
                 verbosity.Equals(VerbosityOptions.d) ||
                 verbosity.Equals(VerbosityOptions.detailed);
         }
+
+        internal static IEnumerable<string> ResolveArchOptionToRuntimeIdentifier(string arg, ParseResult parseResult)
+        {
+            if (parseResult.HasOption(RuntimeOption(string.Empty).Aliases.First()))
+            {
+                throw new GracefulException(CommonLocalizableStrings.CannotSpecifyBothRuntimeAndArchOptions);
+            }
+
+            if (parseResult.BothArchAndOsOptionsSpecified())
+            {
+                // ResolveOsOptionToRuntimeIdentifier handles resolving the RID when both arch and os are specified
+                return Array.Empty<string>();
+            }
+
+            return ResolveRidShorthandOptions(null, arg);
+        }
+
+        internal static IEnumerable<string> ResolveOsOptionToRuntimeIdentifier(string arg, ParseResult parseResult)
+        {
+            if (parseResult.HasOption(RuntimeOption(string.Empty).Aliases.First()))
+            {
+                throw new GracefulException(CommonLocalizableStrings.CannotSpecifyBothRuntimeAndOsOptions);
+            }
+
+            if (parseResult.BothArchAndOsOptionsSpecified())
+            {
+                return ResolveRidShorthandOptions(arg, parseResult.ValueForOption<string>(CommonOptions.ArchitectureOption().Aliases.First()));
+            }
+
+            return ResolveRidShorthandOptions(arg, null);
+        }
+
+        private static IEnumerable<string> ResolveRidShorthandOptions(string os, string arch)
+        {
+            var properties = new string[] { $"-property:RuntimeIdentifier={ResolveRidShorthandOptionsToRuntimeIdentifier(os, arch)}" };
+            return properties;
+        }
+
+        internal static string ResolveRidShorthandOptionsToRuntimeIdentifier(string os, string arch)
+        {
+            var currentRid = GetCurrentRuntimeId();
+            os = string.IsNullOrEmpty(os) ? GetOsFromRid(currentRid) : os;
+            arch = string.IsNullOrEmpty(arch) ? GetArchFromRid(currentRid) : arch;
+            return $"{os}-{arch}";
+        }
+
+        private static string GetCurrentRuntimeId()
+        {
+            string runtimeIdentifierChainPath;
+            var ridFileName = "NETCoreSdkRuntimeIdentifierChain.txt";
+            if (Path.GetFileName(Environment.ProcessPath).Equals("dotnet.exe"))
+            {
```

Comment:
* This will only work on Windows
* Why not do it the other way round - single out the testing scenario - I understand why we want test-only behavior in the product, and I think it's fine to have that. But I think the code should look something like:
```C#
if (Path.GetFileNameWithoutExtension(Environment.ProcessPath).Equals("testhost")
{
    // Do test specific discovery
}
else 
{
   // Do normal product code path
}
```

I think it's better to be very specific about supporting tests. One other way (even cleaner) would be to have for example an environment variable which has to be set in order to activate the test only behavior.

The end goal being that it should be basically impossible to get the test behavior by accident.

Reply:
Investigating the CI failures I realized that we only see the test host being run when we're running with VS, in CI tests are run with dotnet.exe but we still have the same issues, so I don't think it'll be possible to completely separate out this logic.

Reply:
Yeah - I was looking at it as well.. The test stuff is somewhat of a mess. I guess it will have to be "not nice" :-). Please add a comment why it's doing what it's doing and what is the expected environment - just so that a year from now this doesn't look like dark magic.

And thanks for bearing with me 👼

Reply:
Just pushed the change